cirq/protocols/decompose.py

```diff
@@ -0,0 +1,234 @@
+            "acceptable to keep.")
+
+    if intercepting_decomposer is None:
+        actual_decomposer = decomposer
+    else:
+        actual_decomposer = lambda op: _intercept_decompose(
+            op, [intercepting_decomposer, decomposer])
+
+    output = []
+    queue = []  # type: List['cirq.Operation']
+    if not isinstance(val, ops.Operation):
+        queue.extend(decompose_once(val))
+    else:
+        queue.append(val)
+
+    while queue:
+        item = queue.pop(0)
+
+        if keep is not None and keep(item):
+            output.append(item)
+            continue
+
+        decomposed = actual_decomposer(item)
+        if decomposed is not NotImplemented:
+            queue.extend(ops.flatten_op_tree(decomposed))
+            continue
+
+        if isinstance(item, collections.Iterable):
+            queue.extend(ops.flatten_op_tree(item))
+            continue
+
+        if on_stuck_raise is not None:
+            if isinstance(on_stuck_raise, Exception):
+                raise on_stuck_raise
+            else:
+                raise on_stuck_raise(item)
+
+        output.append(item)
+
+    return output
+
+
+def decompose_once(val: Any,
+                   default: TDefault=RaiseTypeErrorIfNotProvided,
+                   **kwargs
+                   ) -> Union[TDefault, List['cirq.Operation']]:
+    """Decomposes a value into operations, if possible.
+
+    Args:
+        val: The value to call `_decompose_` on, if possible.
+        default: A default result to use if the value doesn't have a
+            `_decompose_` method or that method returns `NotImplemented`.
+            If not specified, undecomposable values cause a `TypeError`.
+        kwargs: Arguments to forward into the `_decompose_` method of `val`.
+            For example, this is used to tell gates what qubits they are being
+            applied to.
+
+    Returns:
+        The result of `val._decompose_(**kwargs)`, if `val` has a `_decompose_`
+        method and it didn't return `NotImplemented`. Otherwise `default` is
+        returned, if it was specified. Otherwise an error is raised.
+
+    TypeError:
+        `val` didn't have a `_decompose_` method (or that method returned
+        `NotImplemented`) and `default` wasn't set.
+    """
+    from cirq import ops  # HACK: Avoids circular dependencies.
+
+    method = getattr(val, '_decompose_', None)
+    decomposed = NotImplemented if method is None else method(**kwargs)
+
+    if decomposed is not NotImplemented:
+        return list(ops.flatten_op_tree(decomposed))
+
+    if default is not RaiseTypeErrorIfNotProvided:
+        return default
+    if method is None:
+        raise TypeError("object of type '{}' "
+                        "has no _decompose_ method.".format(type(val)))
+    raise TypeError("object of type '{}' does have a _decompose_ method, "
+                    "but it returned NotImplemented.".format(type(val)))
+
+
+def decompose_once_with_qubits(
```

Comment:
could both decompose_once and decompose take in an optional Qubits for this case?

Reply:
I tried doing that initially. The code looked gross. This way is clearer and makes it harder for the caller to make a mistake such as thinking they need to pass in `operation.qubits` in order to decompose an operation.